numpy/_core/include/numpy/ndarraytypes.h

```diff
@@ -69,13 +69,20 @@ enum NPY_TYPES {    NPY_BOOL=0,
                      */
                     NPY_DATETIME, NPY_TIMEDELTA, NPY_HALF,
 
+                    /* The number of builtin legacy types */
                     NPY_NTYPES,
                     NPY_NOTYPE,
+
+                    /*
+                     * New types added after NumPy 2.0
+                     */
+                    NPY_VSTRING,
```

Comment:
My first attempt at this tried changing NPY_NTYPES and that broke *a ton* of stuff. It looks like there are downstream usages too.

Should I perhaps define `NPY_LEGACY_NTYPES` and deprecate `NPY_NTYPES`? Or should we try changing the value of NPY_NTYPES and define `NPY_LEGACY_NTYPES` for backward compatibility?

Reply:
Not sure how we use NTYPES everywhere.  For downstream it  might have to be runtime dependent.
I do wonder a bit if we should care, I am not sure there is much downstream C-code that will work for the new string dtype (after all it is more complicated than most) without being changed.  At which point, the question is whether avoiding type numbers entirely is a huge hassle.

Or, we just assign a user defined dtype type number (we can even still do that for extension dtypes in principle, at least so long as they have a singleton default instance).

Reply:
Making it a runtime constant seems to work OK and not be problematic elsewhere in NumPy, so long as I take care to make sure `NPY_NOTYPE` can't be stepped on by a builtin dtype number. I also added `NPY_NTYPES_LEGACY` since I suspect that will be useful for downstream in cases where working with new dtypes won't make sense for code that expects a legacy dtype.

Reply:
For what it's worth, `NPY_ABI_COMPATIBLE` remains 21 after this change.